trie-db/src/iterator.rs

```diff
@@ -288,35 +288,60 @@ impl<L: TrieLayout> TrieDBRawIterator<L> {
 		prefix: &[u8],
 		seek: &[u8],
 	) -> Result<(), TrieHash<L>, CError<L>> {
-		if seek.starts_with(prefix) {
-			self.seek(db, seek)?;
-			let prefix_len = prefix.len() * crate::nibble::nibble_ops::NIBBLE_PER_BYTE;
-			let mut len = 0;
-			// look first prefix in trail
-			for i in 0..self.trail.len() {
-				match self.trail[i].node.node_plan() {
-					NodePlan::Empty => {},
-					NodePlan::Branch { .. } => {
-						len += 1;
-					},
-					NodePlan::Leaf { partial, .. } => {
-						len += partial.len();
-					},
-					NodePlan::Extension { partial, .. } => {
-						len += partial.len();
-					},
-					NodePlan::NibbledBranch { partial, .. } => {
-						len += 1;
-						len += partial.len();
-					},
-				}
-				if len > prefix_len {
-					self.trail = self.trail.split_off(i);
-					return Ok(())
-				}
+		if prefix.is_empty() {
+			// There's no prefix, so just seek.
+			return self.seek(db, seek).map(|_| ())
+		}
+
+		if seek.is_empty() || seek <= prefix {
+			// Either we're not supposed to seek anywhere,
+			// or we're supposed to seek *before* the prefix,
+			// so just directly go to the prefix.
+			return self.prefix(db, prefix)
+		}
+
+		if !seek.starts_with(prefix) {
+			// We're supposed to seek *after* the prefix,
+			// so just return an empty iterator.
+			self.trail.clear();
+			return Ok(())
+		}
```

Comment:
This looks like something that shouldn't be supported at all and the upperlying logic should forbid this?

Reply:
> I was thinking about it from the sp_io perspective and api should be I think:
> 
>     * prefix some prefix
> 
>     * seek some seek key without the prefix
> 
> 
> So the first value you access will not be `seek` when into the prefix, but `prefix ++ seek`

So if I understand this right, are you suggesting that the subsequent seek after seeking to the prefix should be allowed to "break out" of the prefix and iterate over values which do not start with the prefix?

Reply:
It is more an api suggestion that would avoid some corner cases (current code looks plain good to me):
Instead of calling `prefix_then_seek(b"prefix", b"prefix_seek")` or `prefix_then_seek(b"prefix", b"not_in_prefix")`.
We will call `prefix_then_seek(b"prefix", b"_seek")` and could not do the second call at all.

Reply:
But doing this change is breaking api. I think it would be more interesting at sp_io level also.
(was just a thinking I had during lunch and wanted to share).

Reply:
Hmmm... I guess we could do that, but that might be more awkward to use depending on the use case I guess? But then it could be easier in some other cases? And also, in this case I think it could make sense for it to return keys which are also have the prefix stripped, e.g. assume the following code:

```rust
loop {
    let accounts = fetch_values(ACCOUNTS_PREFIX, cursor, 100);
    // ...
    cursor = match accounts.last() {
        Some(account) => account,
        None => break
    }
}
```

For this code to work and be convenient to use it'll have to return the same "type" of key (stripped of the prefix or not stripped) as what the function accepts in `cursor`, otherwise you'll have to do awkward concats or have to slice off the prefix yourself.

---

Also, only tangentially related, but maybe naming wise we could rename `prefix` to `only_starting_with` and `seek` to `skip_until_or_after` or something like that? It's a little wordy but it would explain what it does more clearly.

Reply:
Thinking twice could be good for trie to have better naming, but I got pretty use to the current one so my opinion is rather biased.